File: ports/dotnet/src/RussianDocs.Service/Logging/LogRing.cs
using System.Globalization;
using System.Text;
using System.Text.Json;
using System.Text.Json.Serialization;
using Microsoft.Extensions.Logging;

namespace RussianDocs.Service.Logging;

/// <summary>One buffered record. The JSON shape is what the logs page reads.</summary>
public sealed record LogEntry
{
    [JsonPropertyName("ts")] public double Ts { get; init; }
    [JsonPropertyName("level")] public required string Level { get; init; }
    [JsonPropertyName("logger")] public required string Logger { get; init; }
    [JsonPropertyName("message")] public required string Message { get; init; }
    [JsonPropertyName("exc")] public string? Exc { get; init; }
}

/// <summary>
/// The in-memory log buffer behind <c>GET /logs</c>, plus the stdout writer.
///
/// <para>
/// **Two sinks, and the split matters**: stdout carries structured lines at the configured level,
/// while the ring buffer captures EVERYTHING regardless of that level. The reason is the operator
/// workflow — when something goes wrong you want the debug lines that were already emitted, and
/// raising the level afterwards cannot retrieve them.
/// </para>
///
/// <para>Port of <c>service/core/logging.py</c>.</para>
/// </summary>
public static class LogRing
{
    /// <summary>
    /// How many entries the ring holds.
    ///
    /// <para>
    /// 5000, matching the reference, so both services keep the same amount of history and an operator
    /// comparing them is not misled by one having forgotten more. Bounded on purpose: this is an
    /// in-memory diagnostic aid, not a log store, and an unbounded buffer in a long-running service is
    /// a slow leak nobody planned. At roughly 150 bytes per entry that is under a megabyte.
    /// </para>
    /// </summary>
    public const int Capacity = 5000;

    private static readonly Dictionary<string, int> LevelOrder = new(StringComparer.Ordinal)
    {
        ["DEBUG"] = 0, ["INFO"] = 1, ["WARN"] = 2, ["WARNING"] = 2, ["ERROR"] = 3,
        ["CRITICAL"] = 4,
    };

    // An array used circularly rather than a queue: fixed allocation, no per-entry garbage, and the
    // read path is a single pass.
    private static readonly object Gate = new();
    private static readonly LogEntry?[] Entries = new LogEntry?[Capacity];
    private static int _next;
    private static bool _filled;

    internal static void Add(LogEntry entry)
    {
        lock (Gate)
        {
            Entries[_next] = entry;
            _next = (_next + 1) % Entries.Length;
            if (_next == 0)
            {
                _filled = true;
            }
        }
    }

    /// <summary>Entries NEWEST FIRST.</summary>
    private static List<LogEntry> Snapshot()
    {
        lock (Gate)
        {
            int count = _filled ? Entries.Length : _next;
            var output = new List<LogEntry>(count);
            for (int i = 0; i < count; i++)
            {
                int index = (_next - 1 - i + Entries.Length * 2) % Entries.Length;
                if (Entries[index] is { } entry)
                {
                    output.Add(entry);
                }
            }
            return output;
        }
    }

    /// <summary>
    /// The most recent entries, optionally filtered.
    ///
    /// <para>
    /// <paramref name="level"/> is a MINIMUM severity, not an exact match: asking for warnings should
    /// show errors too, which is what an operator means by "show me warnings".
    /// </para>
    /// </summary>
    public static List<LogEntry> Recent(int n, string level, string search)
    {
        int floor = LevelOrder.GetValueOrDefault(level.ToUpperInvariant(), 0);
        string needle = search.ToLowerInvariant();

        var output = new List<LogEntry>(n);
        foreach (LogEntry entry in Snapshot())
        {
            if (LevelOrder.TryGetValue(entry.Level, out int rank) && rank < floor)
            {
                continue;
            }
            if (needle.Length > 0 &&
                !entry.Message.ToLowerInvariant().Contains(needle, StringComparison.Ordinal))
            {
                continue;
            }
            output.Add(entry);
            if (output.Count >= n)
            {
                break;
            }
        }
        return output;
    }

    public static LogLevel ParseLevel(string name) => name.Trim().ToUpperInvariant() switch
    {
        "DEBUG" => LogLevel.Debug,
        "WARNING" or "WARN" => LogLevel.Warning,
        "ERROR" => LogLevel.Error,
        "CRITICAL" => LogLevel.Critical,
        _ => LogLevel.Information,
    };

    /// <summary>The names the Python service uses, so one log pipeline can ingest either.</summary>
    public static string LevelName(LogLevel level) => level switch
    {
        LogLevel.Trace or LogLevel.Debug => "DEBUG",
        LogLevel.Information => "INFO",
        LogLevel.Warning => "WARNING",
        LogLevel.Error => "ERROR",
        _ => "CRITICAL",
    };
}

/// <summary>
/// The logger provider: writes one JSON object per line to stdout, and every record to the ring.
///
/// <para>
/// Hand-written rather than a logging package, for the same reason the JWT is: it is forty lines, and
/// it keeps the service's dependency list at what the library genuinely needs. It also makes the
/// two-sink rule visible in one place instead of spread across provider configuration.
/// </para>
/// </summary>
public sealed class RingLoggerProvider(LogLevel stdoutLevel) : ILoggerProvider
{
    /// <summary>
    /// Serialised because two threads writing partial lines to the same stream interleave them, and
    /// an interleaved JSON line is worse than a dropped one — a log parser rejects the whole file.
    /// </summary>
    private static readonly object WriteGate = new();

    public ILogger CreateLogger(string categoryName) =>
        new RingLogger(categoryName, stdoutLevel, WriteGate);

    public void Dispose() { }

    private sealed class RingLogger(string category, LogLevel stdoutLevel, object writeGate)
        : ILogger
    {
        /// <summary>
        /// Unconditionally TRUE: the buffer wants every level. The stdout level is applied in
        /// <see cref="Log{TState}"/>, so stdout stays at the configured verbosity while the ring
        /// keeps everything.
        /// </summary>
        public bool IsEnabled(LogLevel level) => level != LogLevel.None;

        public IDisposable? BeginScope<TState>(TState state) where TState : notnull => null;

        public void Log<TState>(LogLevel level, EventId eventId, TState state, Exception? error,
            Func<TState, Exception?, string> formatter)
        {
            string message = formatter(state, error);
            string levelName = LogRing.LevelName(level);
            LogRing.Add(new LogEntry
            {
                Ts = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds() / 1000.0,
                Level = levelName,
                Logger = category,
                Message = message,
                Exc = error?.ToString(),
            });

            if (level < stdoutLevel)
            {
                return;
            }

            var line = new StringBuilder();
            line.Append('{');
            line.Append("\"timestamp\":").Append(JsonSerializer.Serialize(
                // Quoted T and Z, for the reason spelled out on NullableUtcConverter.Pattern.
                DateTime.UtcNow.ToString("yyyy-MM-dd'T'HH:mm:ss'Z'",
                    CultureInfo.InvariantCulture)));
            line.Append(",\"level\":").Append(JsonSerializer.Serialize(levelName));
            line.Append(",\"logger\":").Append(JsonSerializer.Serialize(category));
            line.Append(",\"message\":").Append(JsonSerializer.Serialize(message));
            if (error is not null)
            {
                line.Append(",\"exc\":").Append(JsonSerializer.Serialize(error.ToString()));
            }
            line.Append('}');

            lock (writeGate)
            {
                Console.Out.WriteLine(line.ToString());
            }
        }
    }
}

File: ports/java/service/src/main/kotlin/net/russiandocs/service/logging/LogRing.kt
package net.russiandocs.service.logging

import java.io.FileDescriptor
import java.io.FileOutputStream
import java.io.PrintStream
import java.time.Instant
import java.time.ZoneOffset
import java.time.format.DateTimeFormatter
import java.util.Locale
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/** One buffered record. The JSON shape is what the logs page reads. */
@Serializable
public data class LogEntry(
    @SerialName("ts") val ts: Double,
    @SerialName("level") val level: String,
    @SerialName("logger") val logger: String,
    @SerialName("message") val message: String,
    @SerialName("exc") val exc: String? = null,
)

/**
 * The in-memory log buffer behind `GET /logs`, plus the stdout writer.
 *
 * **Two sinks, and the split matters**: stdout carries structured lines at the configured level, while the
 * ring buffer captures EVERYTHING regardless of that level. The reason is the operator workflow — when
 * something goes wrong you want the debug lines that were already emitted, and raising the level afterwards
 * cannot retrieve them.
 *
 * Hand-written rather than routed through SLF4J/Logback, for the same reason the JWT is hand-rolled: the
 * ring buffer is the deliverable here, an appender wrapping it would be more configuration than code, and
 * the two-sink rule stays visible in one file instead of spread across a logback.xml. Spring's own startup
 * lines still go through Logback and look different — that is why `logging.level.root` is turned down in
 * `application.properties`, so the service's own JSON stream is the log an operator reads.
 *
 * Port of `service/core/logging.py`.
 */
public object LogRing {

    /**
     * How many entries the ring holds.
     *
     * 5000, matching the reference, so all implementations keep the same amount of history and an operator
     * comparing them is not misled by one having forgotten more. Bounded on purpose: this is an in-memory
     * diagnostic aid, not a log store, and an unbounded buffer in a long-running service is a slow leak
     * nobody planned. At roughly 150 bytes per entry that is under a megabyte.
     */
    public const val CAPACITY: Int = 5000

    public const val DEBUG: Int = 0
    public const val INFO: Int = 1
    public const val WARNING: Int = 2
    public const val ERROR: Int = 3
    public const val CRITICAL: Int = 4

    private val LEVEL_ORDER: Map<String, Int> = mapOf(
        "DEBUG" to DEBUG, "INFO" to INFO, "WARN" to WARNING, "WARNING" to WARNING,
        "ERROR" to ERROR, "CRITICAL" to CRITICAL,
    )

    // An array used circularly rather than a queue: fixed allocation, no per-entry garbage, and the read
    // path is a single pass.
    private val gate = Any()
    private val entries = arrayOfNulls<LogEntry>(CAPACITY)
    private var next = 0
    private var filled = false

    /**
     * The stdout stream, forced to UTF-8.
     *
     * `System.out` on Windows uses the console codepage, which turns every Cyrillic character in a log
     * message into `?`. The conformance CLI hit exactly this and needs the same fix — DEVIATIONS J-10.
     */
    private val stdout = PrintStream(FileOutputStream(FileDescriptor.out), true, "UTF-8")

    private val TIMESTAMP: DateTimeFormatter =
        DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ss'Z'").withZone(ZoneOffset.UTC)

    /** The minimum severity written to stdout. The ring ignores it — see the type note. */
    @Volatile
    public var stdoutLevel: Int = INFO

    public fun add(entry: LogEntry): Unit = synchronized(gate) {
        entries[next] = entry
        next = (next + 1) % entries.size
        if (next == 0) {
            filled = true
        }
    }

    /**
     * Records one line: into the ring always, to stdout when it clears [stdoutLevel].
     *
     * The stdout write is inside the same lock as the ring update, because two threads writing partial
     * lines to one stream interleave them — and an interleaved JSON line is worse than a dropped one, since
     * a log parser rejects the whole file.
     */
    public fun log(level: Int, logger: String, message: String, error: Throwable? = null) {
        val name = levelName(level)
        val entry = LogEntry(
            ts = System.currentTimeMillis() / 1000.0,
            level = name,
            logger = logger,
            message = message,
            exc = error?.stackTraceToString(),
        )
        add(entry)
        if (level < stdoutLevel) {
            return
        }
        val line = StringBuilder()
        line.append('{')
        line.append("\"timestamp\":").append(quote(TIMESTAMP.format(Instant.now())))
        line.append(",\"level\":").append(quote(name))
        line.append(",\"logger\":").append(quote(logger))
        line.append(",\"message\":").append(quote(message))
        if (error != null) {
            line.append(",\"exc\":").append(quote(error.stackTraceToString()))
        }
        line.append('}')
        synchronized(gate) { stdout.println(line) }
    }

    /** Escapes a JSON string. Small enough to write out, and it keeps the log path allocation-light. */
    private fun quote(value: String): String {
        val sb = StringBuilder(value.length + 2)
        sb.append('"')
        for (ch in value) {
            when (ch) {
                '"' -> sb.append("\\\"")
                '\\' -> sb.append("\\\\")
                '\n' -> sb.append("\\n")
                '\r' -> sb.append("\\r")
                '\t' -> sb.append("\\t")
                else -> if (ch < ' ') {
                    sb.append("\\u").append("%04x".format(ch.code))
                } else {
                    sb.append(ch)
                }
            }
        }
        sb.append('"')
        return sb.toString()
    }

    /** Entries NEWEST FIRST. */
    private fun snapshot(): List<LogEntry> = synchronized(gate) {
        val count = if (filled) entries.size else next
        val output = ArrayList<LogEntry>(count)
        for (i in 0 until count) {
            val index = (next - 1 - i + entries.size * 2) % entries.size
            entries[index]?.let { output.add(it) }
        }
        output
    }

    /**
     * The most recent entries, optionally filtered.
     *
     * [level] is a MINIMUM severity, not an exact match: asking for warnings should show errors too, which
     * is what an operator means by "show me warnings".
     */
    public fun recent(n: Int, level: String, search: String): List<LogEntry> {
        val floor = LEVEL_ORDER[level.uppercase(Locale.ROOT)] ?: DEBUG
        val needle = search.lowercase(Locale.ROOT)

        val output = ArrayList<LogEntry>(n)
        for (entry in snapshot()) {
            val rank = LEVEL_ORDER[entry.level]
            if (rank != null && rank < floor) {
                continue
            }
            if (needle.isNotEmpty() && !entry.message.lowercase(Locale.ROOT).contains(needle)) {
                continue
            }
            output.add(entry)
            if (output.size >= n) {
                break
            }
        }
        return output
    }

    public fun parseLevel(name: String): Int = when (name.trim().uppercase(Locale.ROOT)) {
        "DEBUG" -> DEBUG
        "WARNING", "WARN" -> WARNING
        "ERROR" -> ERROR
        "CRITICAL" -> CRITICAL
        else -> INFO
    }

    /** The names the Python service uses, so one log pipeline can ingest either. */
    public fun levelName(level: Int): String = when (level) {
        DEBUG -> "DEBUG"
        WARNING -> "WARNING"
        ERROR -> "ERROR"
        CRITICAL -> "CRITICAL"
        else -> "INFO"
    }
}

/**
 * A named logger over [LogRing].
 *
 * The service passes `::info`-style function references around as `(String) -> Unit` — the store and the
 * repositories take exactly that, so nothing below the API layer knows what logging is.
 */
public class ServiceLog(private val name: String) {
    public fun debug(message: String): Unit = LogRing.log(LogRing.DEBUG, name, message)
    public fun info(message: String): Unit = LogRing.log(LogRing.INFO, name, message)
    public fun warn(message: String): Unit = LogRing.log(LogRing.WARNING, name, message)
    public fun error(message: String, cause: Throwable? = null): Unit =
        LogRing.log(LogRing.ERROR, name, message, cause)

    public fun critical(message: String): Unit = LogRing.log(LogRing.CRITICAL, name, message)

    /** The sink form, for the layers that must not know about logging at all. */
    public fun sink(): (String) -> Unit = ::info
}
